### DISCUSSION TITLE
Gemini supports the OpenAI API now. Did anyone try to access it using simple-openai?

### DISCUSSION THREAD

**Discussion by the-gigi** [sashirestela/simple-openai#221] [Ideas]
See https://developers.googleblog.com/en/gemini-is-now-accessible-from-the-openai-library/

It will be very cool if simple-openai can add Gemini to its supported providers. 

According to the example, all it takes is a different base URL and model name. No complex adaptation like Azure OpenAI.

**Comment by the-gigi** [comment]
I plan to try it in the next couple of weeks. I'll report back here and if it works as expected it can be added to the documentation as yet another supported provider.

**Comment by the-gigi** [comment]
At the moment I'm unable to use simple-openai as is to access Gemini due to BaseSimpleOpenAI constructor being package-private.
See https://github.com/sashirestela/simple-openai/issues/247

**Comment by the-gigi** [comment]
There is one more issue that is different with Gemini. Its API key is a token that you need to refresh  every hour. Currently, when you create an OpenAI instance it takes an API key as a string and uses it for as long as this instance is alive. 

This can be handled by the user by creating a new instance OpenAI instance periodically. But, this is pretty cumbersome and potentially requires a lot of changes in many call sites if you call site just receive an OpenAI instance at initialization and then just use it.

The API key is converted to an auth header passed to CleverClient

I can see several ways to address this concern:

1. add an apiKey setter that will create a new CleverClient, which is identical to the existing one except the auth header contains the new API key (no changes need to CleverClient).
2. similar, but also add updateHeaders method to CleverClient that will allow updating the auth header of the existing CleverClient without re-creating a whole new CleverClient (requires changes to CleverClient
3. Use the existing requestInterceptor (requires no changes to somple-openai or cleverclient)

I think that using the request interceptor is a valid approach unless this use case of a dynamci API key is something that simple-openai should support as first-class citizen.

Note that specifically for Gemini and request interceptor is required regradless because it need to strip the "/v1" version that simple-openai automatically attaches to the base url for some calls. So, extending this request interceptor to handle the auth header too doesn't add a lot of complexity.

@sashirestela WDYT?

**Reply by the-gigi** [reply]
I see you're working on a PR for improving support for multiple providers:
https://github.com/sashirestela/simple-openai/pull/246

Please consider support for dynamic API key 🙏

**Reply by sashirestela** [reply]
@the-gigi Thanks for your research about Gemini!

> 3. Use the existing requestInterceptor (requires no changes to somple-openai or cleverclient)
>
> I think that using the request interceptor is a valid approach unless this use case of a dynamci API key is something that simple-openai should support as first-class citizen.
> 
> Note that specifically for Gemini and request interceptor is required regradless because it need to strip the "/v1" version that simple-openai automatically attaches to the base url for some calls. So, extending this request interceptor to handle the auth header too doesn't add a lot of complexity.

Agree, Request Interceptor is the most appropriate approach for this new provider.

> Please consider support for dynamic API key 🙏

Request Interceptor will be the approach to handle this particular Gemini functionality, no additional support will be included.

**Reply by the-gigi** [reply]
@sashirestela  I was able to implement the support for Gemini in a fork. This includes the dynamic API key handling with no changes to the core classes, so core simple-openai doesn't need to provide special support for dynamic API keys.

The `SimpleOpenAIGemini` class takes `Supplier<String> apiKeyProvider` as an argument, instead of a fixed `String apiKey` like other providers. Then in the requestInterceptor it invokes it to get a fresh token and setting the auth header for the request.

I also implemented `ChatGeminiDemo`. Everything works except for 
```
        /* requires GCS file path. Didn't test */
        //demo.addTitleAction("Call Chat with Vision (External image)", demo::demoCallChatWithVisionExternalImage);
```

This is due to requirement by Gemini to use Google cloud storage files, so the wikipedia URL doesn't work.

See https://github.com/the-gigi/simple-openai/commit/40fe114eefdf83b589abe27f74a4b20dc1a81138

Let me know if you want me to open a PR, or you can just use it as inspiration.

**Comment by sashirestela** [comment]
@the-gigi It is awesome that you have verified the feasibility of the Gemini API with your POC.
- Could you create an issue labelled as enhancement for this, please?
- Any PR for this should be worked after the next release and taking the changes related to providers (I have even more changes in progress). The next release should be delivered today in the next hours.

**Reply by the-gigi** [reply]
Done. https://github.com/sashirestela/simple-openai/issues/248

I don't think I can label issues.

**Reply by the-gigi** [reply]
@sashirestela I rebased my fork on 3.13.0 and updated the `OpenAIGemini`  + `ChatGeminiDemo` classes to the new approach. Let me know if it's OK to open a PR

https://github.com/the-gigi/simple-openai/commit/87cdaaa6a75908fe02e37f72872d4c11865746de

**Reply by sashirestela** [reply]
Please, rebase to [release_3.14.0](https://github.com/sashirestela/simple-openai/tree/release_3.14.0) which has the latest refactor to Demo code structure. Are you sure that Gemini Vertex implemnents all the services? If that is the case, you should add Demo code for each of them. Take as a reference the Mistral implementation demo code.

**Reply by the-gigi** [reply]
Oh, wow. You move fast :-) I'll rebase to 3.14.0 and check out the Mistral demo.

**Reply by the-gigi** [reply]
@sashirestela I rebased my fork on [release_3.14.0](https://github.com/sashirestela/simple-openai/tree/release_3.14.0)

I also simplified it to only support chat completion for now.

https://github.com/sashirestela/simple-openai/compare/main...the-gigi:simple-openai:main

**Comment by the-gigi** [comment]
@sashirestela Cleverclient's 2.x now has response interceptor 💯 . simple-openai currently uses [Cleverclient version 1.6.3](https://github.com/sashirestela/simple-openai/blob/main/pom.xml#L57)

Do you plan to bump to it to latest and expose the response interceptor like the request interceptor here:
https://github.com/sashirestela/simple-openai/blob/main/src/main/java/io/github/sashirestela/openai/base/OpenAIProvider.java#L45?

**Reply by the-gigi** [reply]
actually, I'll just open an issue...

**Reply by the-gigi** [reply]
ha! you just created an issue. What a timing!
